File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/Drivers/BrowserSession.cs
using OpenQA.Selenium;

namespace SauceDemo.UiTests.Drivers;

/// <summary>
/// Owns a WebDriver instance and its dedicated temp profile directory (used to avoid Chrome/Edge's
/// "user data directory is already in use" error when several sessions run concurrently). Disposing
/// quits the browser and best-effort deletes the temp profile; idempotent so a double-dispose from
/// hook ordering mistakes is harmless.
/// </summary>
public sealed class BrowserSession : IDisposable
{
    private readonly string? _userDataDir;
    private bool _disposed;

    public BrowserSession(IWebDriver driver, string? userDataDir)
    {
        Driver = driver;
        _userDataDir = userDataDir;
    }

    public IWebDriver Driver { get; }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;

        try
        {
            Driver.Quit();
        }
        finally
        {
            Driver.Dispose();
            DeleteUserDataDir();
        }
    }

    private void DeleteUserDataDir()
    {
        if (string.IsNullOrEmpty(_userDataDir) || !Directory.Exists(_userDataDir))
        {
            return;
        }

        try
        {
            Directory.Delete(_userDataDir, recursive: true);
        }
        catch (IOException)
        {
            // Best-effort cleanup; the OS temp directory is reclaimed eventually.
        }
        catch (UnauthorizedAccessException)
        {
            // Best-effort cleanup; the OS temp directory is reclaimed eventually.
        }
    }
}

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/drivers/BrowserSession.java
package com.saucedemo.uitests.drivers;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;
import org.openqa.selenium.WebDriver;

/**
 * Owns a WebDriver instance together with the throwaway browser profile directory created for it.
 * Closing quits the browser and then best-effort deletes that directory; both steps are idempotent, so
 * a double close from a hook-ordering mistake is harmless.
 */
public final class BrowserSession implements AutoCloseable {

    private final WebDriver driver;
    private final Path userDataDir;
    private boolean closed;

    BrowserSession(WebDriver driver, Path userDataDir) {
        this.driver = driver;
        this.userDataDir = userDataDir;
    }

    public WebDriver driver() {
        return driver;
    }

    @Override
    public void close() {
        if (closed) {
            return;
        }
        closed = true;

        try {
            driver.quit();
        } finally {
            deleteUserDataDir();
        }
    }

    private void deleteUserDataDir() {
        if (userDataDir == null || !Files.exists(userDataDir)) {
            return;
        }

        try (var paths = Files.walk(userDataDir)) {
            // Deepest-first, because a directory can only be deleted once it is empty.
            paths.sorted(Comparator.reverseOrder()).forEach(path -> {
                try {
                    Files.deleteIfExists(path);
                } catch (IOException ignored) {
                    // Best-effort cleanup; the OS temp directory is reclaimed eventually.
                }
            });
        } catch (IOException ignored) {
            // Same - never let profile cleanup turn into a scenario failure.
        }
    }
}
